include/deal.II/trilinos/nox.h

```diff
@@ -0,0 +1,1029 @@
+
+      /**
+       * Absolute l2 tolerance to be reached.
+       */
+      const double abs_tol;
+
+      /**
+       * Relative l2 tolerance to be reached.
+       */
+      const double rel_tol;
+    };
+
+    /**
+     * Constructor.
+     */
+    NOXSolver(AdditionalData &                            additional_data,
+              const Teuchos::RCP<Teuchos::ParameterList> &parameters);
+
+    /**
+     * Solve non-linear problem.
+     */
+    unsigned int
+    solve(VectorType &solution);
+
+    /**
+     * User function that computes the residual.
+     */
+    std::function<int(const VectorType &x, VectorType &f)> residual = {};
+
+    /**
+     * User function that sets up the Jacobian.
+     */
+    std::function<int(const VectorType &x)> setup_jacobian = {};
+
+    /**
+     * User function that applies the Jacobian.
+     *
+     * @note The function is optional and is used in the case of certain
+     * configurations.
+     */
+    std::function<int(const VectorType &x, VectorType &v)> apply_jacobian = {};
+
+    /**
+     * User function that applies the inverse of the Jacobian.
+     *
+     * @note The function is optional and is used in the case of certain
+     * configurations.
+     */
+    std::function<
+      int(const VectorType &f, VectorType &x, const double tolerance)>
+      solve_with_jacobian = {};
+
+    /**
+     * User function that allows to check convergence in addition to
+     * ones checking the l2-norm and the number of iterations (see
+     * AdditionalData).
+     */
+    std::function<SolverControl::State(const unsigned int,
+                                       const double,
+                                       const VectorType &,
+                                       const VectorType &)>
+      check_iteration_status = {};
+
+  private:
+    /**
+     * Additional data with basic settings.
+     */
+    AdditionalData &additional_data;
+
+    /**
+     * Additional data with advanced settings. An overview of
+     * possible parameters is given at
+     * https://docs.trilinos.org/dev/packages/nox/doc/html/parameters.html.
+     */
+    const Teuchos::RCP<Teuchos::ParameterList> parameters;
+  };
+} // namespace TrilinosWrappers
+
+
+
+#  ifndef DOXYGEN
+
+namespace TrilinosWrappers
+{
+  namespace internal
+  {
+    template <typename VectorType>
+    class Group;
+
+    /**
+     * Implementation of the abstract interface
+     * NOX::Abstract::Vector for deal.II vectors. For details,
+     * see
+     * https://docs.trilinos.org/dev/packages/nox/doc/html/classNOX_1_1Abstract_1_1Vector.html.
+     */
+    template <typename VectorType>
+    class Vector : public NOX::Abstract::Vector
+    {
+    public:
+      /**
+       * Create empty vector.
+       */
+      Vector() = default;
+
+      /**
+       * Wrap an existing vector.
+       */
+      Vector(VectorType &vector)
+      {
+        this->vector.reset(&vector, [](auto *) { /*nothing to do*/ });
+      }
```

Comment:
I don't think we do this for internal classes/functions. Many of the functions are only delegating function calls to our own classes so that the functions are a few lines long (many only one). Splitting the declarations and implementations would just increase the line of numbers without given too much benefit so that I would like to prefer not do this.

Reply:
It's what we do nearly everywhere. Let's try to be consistent about it. (You can even let your IDE do it for you :-) )